src/java/org/apache/cassandra/utils/JVMStabilityInspector.java

```diff
@@ -125,7 +123,31 @@ public static void inspectThrowable(Throwable t, boolean propagateOutOfMemory, C
             killer.killCurrentJVM(t);
 
         if (t.getCause() != null)
-            inspectThrowable(t.getCause(), propagateOutOfMemory, fn);
+            inspectThrowable(t.getCause(), fn);
+    }
+
+    /**
+     * Intentionally produce a heap space OOM upon seeing a Direct buffer memory OOM.
+     * Direct buffer OOM cannot trigger JVM OOM error related options,
+     * e.g. OnOutOfMemoryError, HeapDumpOnOutOfMemoryError, etc.
+     * See CASSANDRA-15214 for more details
+     */
+    @Exclude // Exclude from just in time compilation.
+    private static void forceHeapSpaceOomMaybe(OutOfMemoryError oom)
+    {
+        // See the oom thrown from java.nio.Bits.reserveMemory.
+        if (!"Direct buffer memory".equals(oom.getMessage()))
```

Comment:
Depending on a string from JVM internal libraries seems like it has the potential to break easily in the future. Unless the Error Strings are something the JVM API is committed to keeping (which I don't have any recollection of being the case).

Reply:
Agree. It is a hacky approach.
The message remains the same for jdk 11 and below. 
But the error message has changed since jdk 13. [[1]](https://github.com/openjdk/jdk13u/blob/a0651dd95b40c41803c938781f810aa40c43737b/src/java.base/share/classes/java/nio/Bits.java#L175), [[2]](https://github.com/openjdk/jdk15u/blob/ac465695d5924dc2a634d660c60f2d54d3d2743d/src/java.base/share/classes/java/nio/Bits.java#L175)

I think we can improve it by checking the stack trace that the OOM is thrown from `java.nio.Bits. reserveMemory`. It is more future proofing (valid through jdk15 at least). But still a hack though. 
WDYT?

Reply:
That’s more what I had in mind, yes. Not perfect but more future proof.

Reply:
A downside is that checking trace is more expensive.

I can update the condition to short circuit when the message contains `"direct buffer memory"`